spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -1633,17 +1635,17 @@ private void fixTxOffsetsIfNeeded() {
 			if (this.fixTxOffsets) {
 				try {
 					Map<TopicPartition, OffsetAndMetadata> toFix = new HashMap<>();
-					this.lastCommits.forEach((tp, oamd) -> {
+					this.savedPositions.forEach((tp, saved) -> {
 						long position = this.consumer.position(tp);
-						Long saved = this.savedPositions.get(tp);
+						OffsetAndMetadata comitted = this.lastCommits.get(tp);
```

Comment:
`this.lastCommits` only contains offsets for `TopicPartition`s for which records were obtained during the previous `poll()` call. `this.lastCommits` is cleared between `poll`s.
This means:
* `poll`s returning no records will always trigger a commit for the complete assignment.
* `poll`s returning records for some partitions only will trigger commits for the other partitions even when the LSO did not change since the last commit.

Should the fix try to avoid unnecessary commits or is it good enough as is?

Reply:
@bgK I guess there is a situation where a producer might keep rolling back transactions and the lag will increase until a successful transaction appears in the log. 

I must admit that I am not comfortable doing a commit for each empty poll but if the above is true, then there is probably no choice.

I really wish the kafka folks would fix the underlying problem of reporting this bogus lag.

Reply:
> I must admit that I am not comfortable doing a commit for each empty poll but if the above is true, then there is probably no choice.

Me neither. What I was hinting at with this comment is whether Spring-Kafka should remember the last committed offsets across polls to avoid redundant commits.

Reply:
Right, but as I described, I believe there are cases where the lag can still increase unless we do.

I believe the proper fix is for Kafka to not report a lag if it's a bogus lag due to uncommitted, or rolled-back, records.

Reply:
> Right, but as I described, I believe there are cases where the lag can still increase unless we do.

I agree Spring-Kafka should commit the offsets when the consumer position changed after empty polls. What's unclear to me is what committing the same offsets multiple times when the position does not change after empty polls would accomplish. Do you have any insight?

> I believe the proper fix is for Kafka to not report a lag if it's a bogus lag due to uncommitted, or rolled-back, records.

As far as I know, at this point, Kafka does not provide any other mechanism to report the consumer lag accurately when using transactional producers. Shouldn't Spring-Kafka make sure the situations where Kafka reports the lag properly work when using the Spring abstractions?